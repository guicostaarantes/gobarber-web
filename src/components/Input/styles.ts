import styled, { css } from 'styled-components';

import { CONSTANTS } from '../../styles/constants';

interface ContainerProps {
  isFocused: boolean;
  isFilled: boolean;
  isErrored: boolean;
}

export const Container = styled.div<ContainerProps>`
  background: ${CONSTANTS.grayColor1};
  border-radius: 10px;
  padding: 16px;
  width: 100%;
  display: flex;
  align-items: center;
  margin: 8px;
  color: ${CONSTANTS.grayColor5};
  border: 2px solid ${CONSTANTS.grayColor1};
  transition: color 0.2s, border 0.2s;

  ${(props) =>
    props.isErrored &&
    css`
      border: 2px solid #c53030;
    `}

  ${(props) =>
    props.isFocused &&
    css`
      color: ${CONSTANTS.primaryColor};
      border: 2px solid ${CONSTANTS.primaryColor};
    `}

  ${(props) =>
    props.isFilled &&
    css`
      color: ${CONSTANTS.primaryColor};
    `}

  input {
    background: transparent;
    flex: 1;
    border: 0;
    color: ${CONSTANTS.grayColor8};
    &::placeholder {
      color: ${CONSTANTS.grayColor5};
    }
  }

  svg + input {
    margin-left: 16px;
  }

  input + div {
    margin-left: 16px;
  }
`;
