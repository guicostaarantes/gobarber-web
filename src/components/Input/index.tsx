import React, {
  InputHTMLAttributes,
  useRef,
  useEffect,
  useState,
  useCallback,
} from 'react';

import { useField } from '@unform/core';

import { IconBaseProps } from 'react-icons';
import { FiAlertCircle } from 'react-icons/fi';

import Tooltip from '../Tooltip';

import { Container } from './styles';

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  name: string;
  icon?: React.ComponentType<IconBaseProps>;
}

const Input: React.FC<InputProps> = ({ icon: Icon, name, ...otherProps }) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const { fieldName, defaultValue, registerField, error } = useField(name);
  const [isFocused, setIsFocused] = useState(false);
  const [isFilled, setIsFilled] = useState(false);
  const [isErrored, setIsErrored] = useState(false);

  useEffect(() => {
    registerField({
      name: fieldName,
      ref: inputRef.current,
      path: 'value',
    });
  }, [fieldName, registerField]);

  useEffect(() => {
    setIsErrored(!!error);
  }, [error]);

  const handleFocus = useCallback(() => {
    setIsFocused(true);
  }, []);

  const handleBlur = useCallback(() => {
    setIsFocused(false);
    setIsErrored(false);
    setIsFilled(!!inputRef.current?.value);
  }, []);

  return (
    <Container isFocused={isFocused} isFilled={isFilled} isErrored={isErrored}>
      {Icon && <Icon size={20} />}
      <input
        data-testid="input"
        ref={inputRef}
        defaultValue={defaultValue}
        name={fieldName}
        onFocus={handleFocus}
        onBlur={handleBlur}
        {...otherProps}
      />
      {error && isErrored && (
        <Tooltip icon={FiAlertCircle} text={error} color="#c53030" />
      )}
    </Container>
  );
};

export default Input;
