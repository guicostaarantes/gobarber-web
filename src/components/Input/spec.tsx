import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Form } from '@unform/web';
import { FormHandles } from '@unform/core';
import Input from './index';

describe('Input unit test', () => {
  it('should render correctly', () => {
    const fn = jest.fn();
    const formRef = React.createRef<FormHandles>();
    const { container } = render(
      <Form ref={formRef} onSubmit={fn}>
        <Input name="email" />
      </Form>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should change when typing', async () => {
    const fn = jest.fn();
    const formRef = React.createRef<FormHandles>();
    const { getByTestId } = render(
      <Form ref={formRef} onSubmit={fn}>
        <Input name="email" />
      </Form>,
    );
    const input = getByTestId('input');
    fireEvent.change(input, { target: { value: 'test@test.com' } });
    expect(formRef.current?.getData()).toEqual({ email: 'test@test.com' });
  });
});
