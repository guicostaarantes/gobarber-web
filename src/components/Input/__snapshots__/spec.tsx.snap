// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input unit test should render correctly 1`] = `
.c0 {
  background: #232129;
  border-radius: 10px;
  padding: 16px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #666360;
  border: 2px solid #232129;
  -webkit-transition: color 0.2s,border 0.2s;
  transition: color 0.2s,border 0.2s;
}

.c0 + div {
  margin-top: 8px;
}

.c0 input {
  background: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  color: #f4ede8;
}

.c0 input::-webkit-input-placeholder {
  color: #666360;
}

.c0 input::-moz-placeholder {
  color: #666360;
}

.c0 input:-ms-input-placeholder {
  color: #666360;
}

.c0 input::placeholder {
  color: #666360;
}

.c0 svg + input {
  margin-left: 16px;
}

.c0 input + div {
  margin-left: 16px;
}

<div>
  <form>
    <div
      class="c0"
    >
      <input
        data-testid="input"
        name="email"
        value=""
      />
    </div>
  </form>
</div>
`;
